hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/ozone/Hadoop27RpcTransport.java

```diff
@@ -67,6 +69,12 @@ public OMResponse submitRequest(OMRequest payload) throws IOException {
     try {
       return proxy.submitRequest(NULL_RPC_CONTROLLER, payload);
     } catch (ServiceException e) {
+      OMNotLeaderException notLeaderException =
```

Comment:
Not sure about this one, but not tested. Do we really get a `NotLeaderExecption` if a simple client tries to access an HA OM node? I think there is a chance to connect directly to the leader, and in that case the exception can be different. 

Or am I wrong?


It may be better to throw this exception

Another approach is using  `OmUtils.isOmHAServiceId(conf, omHost)`. For example, we can override `OzoneClientAdapter createAdapter(ConfigurationSource conf, String omHost, int omPort)` in the `ozonefs-hadoop2` version of `RootedOzFs` and  `OzFs`. After checking the HA config whith this utlity we can call the parent implementation (if not HA)

Reply:
@elek Thanks for the comment.

Can reproduce the error by the following steps:
1. Prepare an OM-HA cluster.
2. Prepare the config files in HADOOP_CONF_DIR, including the core-site.xml and the ozone-site.xml.
3. Run the command of "hdfs dfs -ls o3fs://bucket.volume.cluster1/" on a follower OM. The client would get the NotLeaderExecption.
4. Run the command of "hdfs dfs -ls o3fs://bucket.volume.cluster1/" on the leader OM. The client would get normal result.
5. Run the command of "hdfs dfs -ls o3fs://bucket.volume.cluster1/" on other servers. The client would be retring to connect to "0.0.0.0:9862".

When Hadoop27RpcTransport is trying to initialize the socket to OM, client would use the config of "ozone.om.address", and for an OM-HA cluster, this value should be default as "0.0.0.0". 
So when client initialize the proxy on a follower OM host, it will try to connect to the follower OM, and an NotLeaderException will be thrown. And if the client is on the leader server, the proxy would be successfully created and no exceptions should be thrown. And If the client is on other servers, it will try to connect to the local OM, while on OM is running, so the client would be retrying to connect "0.0.0.0:9862".

Reply:
> And if the client is on the leader server, the proxy would be successfully created and no exceptions should be thrown

Thanks the clarification (and sorry for the late answer).

Based on my understanding this code throws exception only if the client is connected to a follower OM. While the client can be used when the client is lucky and connected to the OM.

While I can accept this behavior, it seems to be confusing. One Ozone user may think that hadoop2 compatible Ozone client can be used together with OM HA, but in fact it's unsafe beacuse it can be failed after any of the leader election.

I would consider to always throw an exception when HA is configured (but I can be convinced from the opposite).

What do you think?

Reply:
Thanks for the reply. 
I agree with the solution to always throw an exception since the client can only work when it's on the same server as the leader OM.
I will change the PR later.
Also, I'd like to add HA support for Hadoop 2, hope you can have a review. Thanks a lot.